src/lib.rs

```diff
@@ -1,6 +1,189 @@
 use std::ffi::OsString;
 use std::iter::FromIterator;
 
+/// Structure that represent a unique Variable.
+#[derive(Clone, Debug)]
+pub struct Var {
+    /// The variable's key.
+    key: OsString,
+    /// The variable's value.
+    value: OsString,
+    /// Conditional statement for this variable.
+    condition: bool,
+    /// The environment that own the variable.
+    environment: Environment,
+}
+
+impl Var {
+    /// Check if the Var's value is strictly different of the input.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// extern crate environment;
+    ///
+    /// let var: environment::Var = ("FOO", "BAR").into();
+    ///
+    /// assert_eq!(var.clone().isnt("BAR").assign("BAZ").var("FOO"), var);
+    /// ```
+    pub fn isnt<T: ToString>(mut self, value: T) -> Self {
+        self.condition = self.value != OsString::from(value.to_string());
+
+        self
+    }
+
+    /// Check if the Var's value is strictly equal to the input.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// extern crate environment;
+    ///
+    /// let var: environment::Var = ("FOO", "BAR").into();
+    ///
+    /// assert_eq!(var.clone().is("BAZ").assign("BAZ").var("FOO"), var);
+    /// ```
+    pub fn is<T: ToString>(mut self, value: T) -> Self {
+        self.condition = self.value == OsString::from(value.to_string());
+
+        self
+    }
+
+    /// Assign a new value to the variable.
+    ///
+    /// Note: assigning can be conditioned by
+    /// [is](struct.Var.html#method.is) or [isnt](struct.Var.html#method.isnt)
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// extern crate environment;
+    ///
+    /// let var: environment::Var = ("FOO", "BAR").into();
+    ///
+    /// assert_eq!(var.clone().is("BAZ").assign("BAZ").var("FOO"), var);
+    ///
+    /// assert!(var.clone().isnt("BAZ").assign("BAZ").var("FOO") != var);
+    ///
+    /// ```
+    pub fn assign<T: ToString>(mut self, value: T) -> Environment {
+        if self.condition {
+            self.value = OsString::from(value.to_string());
+        }
+
+        self.environment.vars.push((self.key, self.value));
```

Comment:
With this being outside of the condition, it'll push on a default `value`.  I doubt this is what people expect with the conditions.

It might be good to do the same for the rest.

Reply:
I think this need to be outside because we need to replicate the state of the variable no matter what occurs.

if you do something like:
```rust
("foo", "bar")
    .into() // Create a Var object: state { k: foo, v: bar }
    .isnt("bar") // Conditional action: false
    .assign("baz") // Action hasn't applied and return the inner environment  : state { k: foo, v: bar }
    .var("foo"); // Get the 'foo' var from Environment: state { k: foo, v: }

// you expect to have "bar", without the push outside of the condition you have ""
```

Reply:
Its expected to support converting from a bare `Var` into an `Environment`?  That seems like an odd use case that causes these conflicting surprises in the API
- Push outside of condition and you arbitrarily add blank strings
- Push inside of condition and lose the original value.

Reply:
The Environment outputed by `assign` is the underlying environment for chaining.

Reply:
Sorry, I might not have been clear.

Yes, `assign` returns the chained `Environment` but in the case of `("foo", "bar")`, an `Environment` is implicitly created that is empty, so if the assign fails, `("foo", "bar")` will not be put into the environment. 

In the case of chaining a `Var` off of an `Environment`, this shouldn't be a problem.  For example:
```rust
Environment::with_var(("foo", "bar"))
    .isnt("bar")
    .assign("baz")
   .var("foo");

// you expect "foot=bar" regardless of whether the `push` is inside the condition or not
```

So my question is if `let v: Var = ("foo", "bar").into()` is an expected use case or not.  If it is, then we have to choose between two different surprising behaviors.

- Push outside of condition and you arbitrarily add blank strings
- Push inside of condition and lose the original value.

One way of resolving this conflict is to switch the code from
```rust
impl<T: ToString, Z: ToString> From<(T, Z)> for Var {
    fn from(v: (T, Z)) -> Self {
        Self {
            key: OsString::from(v.0.to_string()),
            value: OsString::from(v.1.to_string()),
            condition: true,
            environment: Environment::empty(),
        }
    }
}
```
to
```rust
impl<T: ToString, Z: ToString> From<(T, Z)> for Var {
    fn from(v: (T, Z)) -> Self {
        let e = Environment::default();
        let e.insert(v)
        e.var(v.0)
    }
}
```
or whatever form works best.  This ensures that `v` is already added to `Environment` so that if a condition fails, it isn't lost.